### ISSUE TITLE
About "no padding" and input size

### ISSUE THREAD

**Issue by cxmscb** [facebookresearch/silk#3]
Thanks for the great work. Currently I would like to train with my own dataset. And I am confused about some implementations.

1.How to implement the "no padding" for the convonlution with kennel 3?
2.Is the input size in training 640*480? And what's the relationship between the size [164, 164] # VGG-4 (146 + 2x9) with the input size?
3.It seems the coco dataset loader codes are missing.

**Comment by gleize** [comment]
Hi @cxmscb,

> 1.How to implement the "no padding" for the convonlution with kennel 3 ?

You can set the `padding` argument to value 0 (either in [Conv2d](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html) or [conv2d](https://pytorch.org/docs/stable/generated/torch.nn.functional.conv2d.html)).

If you want to use our default architecture (VGG-4), it's already using no padding by default (see [here](https://github.com/facebookresearch/silk/blob/007d8584caebae1bcfc75eb2f273e971d751485e/etc/models/silk-vgg.yaml) for the backbone selection).

> 2.Is the input size in training 640*480? And what's the relationship between the size [164, 164] # VGG-4 (146 + 2x9) with the input size?

The input size isn't fixed and varies as the dataset changes.
`[164, 164]` is the size of the crop extracted randomly from the input image.
As you can see [here](https://github.com/facebookresearch/silk/blob/007d8584caebae1bcfc75eb2f273e971d751485e/etc/mode/train-silk.yaml#L89), this crop size is an argument of the [RandomCrop](https://pytorch.org/vision/main/generated/torchvision.transforms.RandomCrop.html) transform specified above it.

> 3.It seems the coco dataset loader codes are missing.

We did not implement a coco dataset class since it already exists [here](https://pytorch.org/vision/main/_modules/torchvision/datasets/coco.html).
We directly instantiate that loader [here](https://github.com/facebookresearch/silk/blob/007d8584caebae1bcfc75eb2f273e971d751485e/etc/datasets/coco/defaults.yaml) by using [Hydra](https://hydra.cc/docs/intro/)'s auto-instantiation of python objects (c.f. [here](https://hydra.cc/docs/advanced/instantiate_objects/overview/#instantiation-of-builtins)).

**Comment by cxmscb** [comment]
Thanks for the help. The codes are a little bit complicated for me. And for the input size, what is the meaning of 146 + 2x9? It means using conv2d with kennel 3 for 9 times?

**Comment by gleize** [comment]
> Thanks for the help. The codes are a little bit complicated for me.

I definitely agree the code is a bit complicated. This released code is actually a small part of a larger codebase, which was trimmed for public release. So there is a bit of a learning curve to make changes to the codebase.

Don't hesitate to ask as many questions as you need to, and I'll help.

> And for the input size, what is the meaning of 146 + 2x9? It means using "no padding" conv2d with kennel 3 for 9 times?

Yes. The VGG-4 backbone has 8 3x3 unpadded convolutions, and there is one additional unpadded 3x3 convolution in the descriptor head.
This calculation was required to keep the size of the descriptor map constant (at 146x146) when changing the backbone.
